[samuelthen] iP

samuelthen opened the pull request:
# 🌟 Bong: Your Ultimate To-Do List Manager

> **"Your mind is for having ideas, not holding them."**  
> – _David Allen_ ([source](https://dansilvestre.com/productivity-quotes))

**Bong** frees your mind from having to remember things you need to do. It's designed to be:

- **Text-based** 📝
- _Easy to learn_ 🚀
- **FAST** – Super Fast! ⚡

## 🚀 Getting Started with Bong

To get started, just follow these simple steps:

1. **Download Bong** from [this link](https://github.com/nus-cs2103-AY2425S1/ip/pull/522).
2. **Double-click** the downloaded file.
3. **Add your tasks** and let Bong manage them for you.
4. Sit back and relax – **It's FREE!** 🎉

## 🔧 Features

Bong helps you manage your tasks efficiently. Here's what's currently available and what's coming soon:

- [x] **Managing tasks**: Add, remove, or update tasks effortlessly.
- [x] **Managing deadlines**: Add, remove, or update tasks effortlessly.
- [ ] **Reminders** _(coming soon)_ ⏰

## ✨ Example Usage

Here's how you can add a task in Bong:

`todo Finish project report`

## 🎁 Why Choose Bong?

It's FREE! 💰
Designed to be fast and easy to use.
It's your perfect companion to help manage your tasks efficiently!
Give it a try and experience the magic of stress-free task management!

samuelthen committed:
Import base branch for students to start off

As student may not have learnt about branching, let's include a copy of 
the base branch into the main repository.

samuelthen committed:
Initial commit

samuelthen committed:
.gitignore: add *.iml

samuelthen committed:
Move Duke.java out of the package

samuelthen committed:
Tweak tutorial text

samuelthen committed:
README: use numbered list for steps

samuelthen committed:
Fix typo in link to checkstyle config files

samuelthen committed:
Update list of contributors

samuelthen committed:
Add a tutorial on text UI testing

samuelthen committed:
textUiTestingTutorial.md: add missing <br>

samuelthen committed:
textUiTestingTutorial.md: mention updating java/javac commands

samuelthen committed:
Specify checkstyle version

Not specifying the checkstyle version causes it to fail with a cryptic
error.

samuelthen committed:
Adapt JavaFX tutorial for Duke

Rather than to have distinct tutorials building up to a Duke-like
application, let's merge the tutorials into Duke's tutorials
to enhance cohesiveness in the course material.

We also merge the gradle/gradleless tutorials into one by providing a
universal entry point to JavaFX in the style of AddressBook
applications.

Fix header levels in JavaFX Tutorial 1.

Change code samples to use Duke.

Add hints on required import statements.

Add location hints to code snippets to help students find where to copy
and paste them.

Remove nitpicks to make the development process smoother.

Fix usage of `Collections` to `FXCollections`.

Replace image for JavaFX tutorial 3.

Specify location to place images.

Replace a screenshot that referred to the outdated package structure.

Remove reference to DukeStub.

samuelthen committed:
Merge pull request #3 from j-lum/inline-javafx-tutorials

Inline JavaFX tutorials

samuelthen committed:
Update Gradle tutorial to reflect new checkstyle config

samuelthen committed:
Fix typo referring to `HelloWorld`

samuelthen committed:
Add images to JavaFX tutorial 3

samuelthen committed:
Merge pull request #7 from j-lum/tutorial-fix

Address issues #5 and #6

samuelthen committed:
Remove references to DukeStub

samuelthen committed:
Merge pull request #8 from j-lum/tutorial-fix

Remove references to DukeStub

samuelthen committed:
Change indentation in FXML samples to match AB-3

samuelthen committed:
Merge pull request #14 from j-lum/fxml-indent

Change indentation in FXML samples to match AB-3

samuelthen committed:
JavaFX tutorial: Support cross-platform JARs

The OpenJFX plugin expects applications to be modular and bundled
with jlink, resulting in fat jars that are not cross-platform. Let's
manually include the required dependencies so that shadow can package
them properly.

samuelthen committed:
Merge pull request #35 from j-lum/xplatformjfx

JavaFX tutorial: Support cross-platform JARs

samuelthen committed:
Remove unnecessary override annotation

Remove extra '@Override' at the start of the code example.

samuelthen committed:
Merge pull request #41 from jiachen247/patch-2

Remove unnecessary override annotation

samuelthen committed:
Remove generic tutorials

Let's remove generic tutorials from this repo as they are not specific
to this project tempalte

samuelthen committed:
Add text-ui-test files

samuelthen committed:
Add Gradle support

samuelthen committed:
README.md: Fix typo of => or

samuelthen committed:
Improve runtest.* scripts

* Improve the javac command to work for multiple java files.
* In runtest.bat, delete ACTUAL.TXT only if it exists.

samuelthen committed:
README: Update setup instructions to match the latest IDE version

samuelthen committed:
README.md: Tweak template text

* Let's replace 'Feature 1' with 'Feature-ABC'
  reason: some students kept the 'Feature 1' in the final UG.
* Let's tweak the template for features and usage
  reasons: to fix grammar, to distinguish between output and outcome

samuelthen committed:
README: Mention the project language level setting

samuelthen committed:
.gitignore: Fix ACTUAL.txt -> ACTUAL.TXT  (#73)

text-ui-test/runtest.bat|sh scripts generate a file ACTUAL.TXT.
However, .gitignore uses ACTUAL.txt, which means the generated
file will not be ignored by Git on non-Windows OS.

Let's update .gitignore as ACTUAL.txt -> ACTUAL.TXT

samuelthen committed:
Bump gradle and lib version

samuelthen committed:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

samuelthen committed:
README: Update the JDK version 11 -> 17

samuelthen committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

samuelthen committed:
Rename chatbot to Bong and implement greet and exit

samuelthen committed:
Implement echo function

samuelthen committed:
Implement Add and List

samuelthen committed:
Implement Mark as Done

samuelthen committed:
Implement ToDos, Events, Deadlines

samuelthen committed:
Implement Text Ui Testing

samuelthen committed:
Implement exception handling

samuelthen committed:
Implement delete task

samuelthen committed:
Update gitignore to ignore data

samuelthen committed:
Implement saving files into the hard disk

samuelthen committed:
Merge branch-Level-7 into master

samuelthen committed:
Implement DateTime instead of string

samuelthen committed:
Merge branch-Level-8 into master

samuelthen committed:
Implement more OOP

samuelthen committed:
Organize into packages

samuelthen committed:
Merge branch 'add-gradle-support'

samuelthen committed:
Implement JUnit tests for TaskList and Parser

samuelthen committed:
Add JavaDoc

samuelthen committed:
Fix coding standard for import statements

samuelthen committed:
Implement find command

samuelthen committed:
Merge branch 'branch-A-JavaDoc'

samuelthen committed:
Resolve merge conflicts

samuelthen committed:
Resolve merge conflicts

Gilbb commented on src/main/java/bong/utils/Parser.java:
```diff
@@ -0,0 +1,70 @@
+        String arguments = parts.length > 1 ? parts[1] : "";
+
+        switch (commandWord) {
+            case "bye" -> {
```
Perhaps the indentation of `case` clause could be updated according to the Java coding standard?
Noticed that the same occurs in `Storage.java`

seah-minlong replied:
Adding on to what Gilbb has mentioned, it seems that the convention for the switch arrow-notation has been agreed to be as such according to this [issue](https://github.com/nus-cs2103-AY2425S1/forum/issues/49):

```
switch (data.charAt(0)) {
case 'T' -> functionT();
case 'E' -> functionE();
case 'D' -> functionD();
default -> functionDef();
}; 
```
Also, this is not a violation but perhaps the `default` case could be wrapped with `{ }` to standardize it? 😄

Gilbb commented on src/main/java/bong/Bong.java:
```diff
@@ -0,0 +1,75 @@
+            try {
+                String userInput = ui.readCommand();
+                Command command = parser.parseCommand(userInput);
+                CommandResult result = command.execute(tasks);
```
Perhaps there can be better name for `execute` method?

Gilbb reviewed (commented):
Overall, the code is easy to read and understand and well structured too! 😄 
Can't find any major Java coding issues except for the indentation of `case` clauses and a small nitpick regarding the method name.
Well done!! 💯

seah-minlong commented on src/main/java/bong/utils/Parser.java:
```diff
@@ -0,0 +1,70 @@
+package bong.utils;
+
+import bong.command.*;
```
Perhaps the imported classes should be listed explicitly as per the coding standard?

seah-minlong commented on src/main/java/bong/utils/Storage.java:
```diff
@@ -0,0 +1,90 @@
+package bong.utils;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+
+import java.util.ArrayList;
+
+import bong.task.Deadline;
+import bong.task.Event;
+import bong.task.Task;
+import bong.task.Todo;
```
I like how you have listed every imported class explicitly as per the coding standard! 💯

seah-minlong commented on src/main/java/bong/command/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+
+        StringBuilder sb = new StringBuilder("Here are the matching tasks in your list:\n");
+        for (int i = 0; i < matchingTasks.size(); i++) {
+            sb.append(String.format("%d.%s %s\n", i + 1, matchingTasks.get(i).getTypeIcon(), matchingTasks.get(i)));
```
Perhaps this line can be wrapped to make it shorter?

seah-minlong reviewed (commented):
Overall, I found your code easy to read and really well-structured! 👍 I only noticed a few minor coding standard violations 😄

W3iZhi commented on src/main/java/bong/command/DeleteCommand.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Represents a command to delete a task from the task list.
+ */
+public class DeleteCommand implements Command {
```
Good naming for the different types of commands

W3iZhi commented on src/main/java/bong/utils/Storage.java:
```diff
@@ -0,0 +1,90 @@
+     * @param list The {@code ArrayList} of tasks to save.
+     * @throws BongException If an error occurs while saving tasks to the file.
+     */
+    public void save(ArrayList<Task> list) throws BongException {
```
I like the simple naming of the save and load functions

W3iZhi commented on src/main/java/bong/task/TaskList.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     * @param task The {@code Task} to add to the list.
+     */
+    public void addTask(Task task) {
```
I like the naming of your different functions

W3iZhi commented on src/main/java/bong/utils/Ui.java:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @param result The {@code CommandResult} object containing feedback for the user.
+     */
+    public void showCommandResult(CommandResult result) {
```
Maybe shorten the name to showResult, just a nitpick

W3iZhi reviewed (commented):
LGTM, just one minor nit

samuelthen committed:
List imported classes

samuelthen committed:
Implement GUI

samuelthen committed:
Merge branch 'branch-Level-10'

samuelthen committed:
Add assertions to critical parts of the codebase

Currently, some sections of the code do not have explicit checks to
ensure assumptions about the state and behavior of the program. This
can lead to potential bugs or unexpected behaviors going undetected,
especially during runtime, and makes debugging more difficult.

Assertions provide a clear and immediate way to catch logical errors
by terminating execution if certain conditions are not met. This helps
in identifying issues early and maintaining code reliability.

Let's add assertions to critical methods and functions where
important assumptions are made about input values, states, and
outputs. This will include checking for null pointers, valid ranges,
and preconditions for method calls.

Assertions are preferable over traditional error handling in these
cases because they are meant for catching programmer errors, rather
than recoverable runtime errors. Moreover, they can be easily
disabled in production environments for performance optimization
while still providing robust error detection during development.

The addition of assertions will improve code quality by preventing
faulty logic from propagating further in the system.

samuelthen committed:
Improve code quality through enhanced abstraction
The codebase currently has tightly coupled components and scattered implementations, leading to redundancy and maintenance challenges.
Increasing abstraction will decouple components, simplify modifications, and enhance flexibility.
Refactor code to identify and abstract common functionalities into reusable components. Replace specific implementations with these abstractions to reduce redundancy and improve consistency.
This approach promotes modularity, making the codebase more maintainable and adaptable for future changes.

samuelthen committed:
Merge pull request #1 from samuelthen/branch-A-Assertions

Add assertions to critical parts of the codebase

samuelthen committed:
Merge branch 'master' into branch-A-CodeQuality

samuelthen committed:
Merge pull request #2 from samuelthen/branch-A-CodeQuality

Improve code quality through enhanced abstraction

samuelthen committed:
Implement list command which tasks are sorted chronologically

samuelthen committed:
Merge branch 'branch-C-Sort'

T1duS commented on src/main/java/bong/Bong.java:
```diff
@@ -0,0 +1,90 @@
+     * @param userInput The user input string.
+     * @return The feedback message to the user after executing the command.
+     */
+    public String getResponse(String userInput) {
```
Nice job implementing the core logic for handling user commands! One minor suggestion would be to include more specific exception handling for commands that might fail due to user input errors. This way, you could provide more detailed feedback to the user instead of the generic 'An error occurred.' message.

T1duS commented on src/main/java/bong/task/Task.java:
```diff
@@ -0,0 +1,140 @@
+            DateTimeFormatter.ofPattern("yyyy-MM-dd"),          // e.g., "2023-12-31"
+            DateTimeFormatter.ofPattern("MMMM d, yyyy"),        // e.g., "December 31, 2023"
+            DateTimeFormatter.ofPattern("MMM d, yyyy")          // e.g., "Dec 31, 2023"
+    };
```
Very cool to have so many different formatters!

T1duS commented on src/main/java/bong/utils/BongException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Represents a custom exception specific to the Bong application.
+ */
+public class BongException extends Exception {
```
It might have been better to create different subclasses for different exceptions.

T1duS commented:
Excellent use of modularization and abstraction. Very good quality OOP code without any obvious mistakes.

T1duS reviewed (commented):
Excellent use of modularization and abstraction. Very good quality OOP code without any obvious mistakes.

samuelthen committed:
Improve DateTime settings

samuelthen committed:
Add checkstyle and improve code style

samuelthen committed:
Modify the index of find command results to match the actual index

samuelthen committed:
Format command result for find command

samuelthen committed:
Edit User Guide

samuelthen committed:
Format User Guide Markdown

samuelthen committed:
Add UI photo
